src/mudlet-lua/lua/GUIUtils.lua

```diff
@@ -1156,3 +1156,158 @@ do
+  [1]  =  11,
+  [2]  =  22,
+  [3]  =  33,
+  [4]  =  44,
+  [5]  =  55,
+  [6]  =  67,
+  [7]  =  78,
+  [8]  =  89,
+  [9]  = 100,
+  [10] = 111,
+  [11] = 122,
+  [12] = 133,
+  [13] = 144,
+  [14] = 155,
+  [15] = 166,
+  [16] = 177,
+  [17] = 188,
+  [18] = 200,
+  [19] = 211,
+  [20] = 222,
+  [21] = 233,
+  [22] = 244,
+  [23] = 255
+}
+
+local ansiPattern = rex.new("\\e\\[([0-9;]+?)m")
+-- function for converting raw ANSI string into something decho can process
+-- bold, italics, underline not currently supported since decho doesn't support them
+function ansi2decho(text)
+  -- match each set of ansi tags, ie [0;36;40m and convert to decho equivalent.
+  -- this works since both ansi colours and echo don't need closing tags and map to each other
+  local result = rex.gsub(text, ansiPattern, function(s)
+    local output = {} -- assemble the output into this table
+
+    local t = string.split(s, ";") -- split the codes into an indexed table
+
+    -- given an xterm256 index, returns an rgb string for decho use
+    local function convertindex(tag)
+      local floor = math.floor
+      -- code from Mudlets own decoding in TBuffer::translateToPlainText
+
+      local rgb
+      if tag < 8 then
+        rgb = colours[tag]
+      elseif tag < 16 then
+        rgb = lightColours[tag-8]
+      elseif tag < 232 then
+        tag = tag - 16 -- because color 1-15 behave like normal ANSI colors
+
+        r = floor(tag / 36)
+        g = floor((tag-(r*36)) / 6)
+        b = floor((tag-(r*36))-(g*6))
+        rgb = {r*51, g*51, b*51}
+      else
+        local component = grayscaleComponents[tag - 232]
+        rgb = {component, component, component}
+      end
+
+      return rgb
+    end
+
+    -- since fg/bg can come in different order and we need them as fg:bg for decho, collect
+    -- the data first, then assemble it in the order we need at the end
+    local fg,bg
+    local i = 1
+    local floor = math.floor
+    local coloursToUse = colours
+    while i <= #t do
+      local code = t[i]
+
+      if code == '0' then -- reset attributes
+        output[#output+1] = "<r>"
+        fg,bg = nil,nil
+        coloursToUse = colours
+      elseif code == "1" then -- light or bold
+        coloursToUse = lightColours
+      elseif code == "22" then -- not light or bold
+        coloursToUse = colours
+      else
+        local layerCode = floor(code / 10)  -- extract the "layer": 3 is fore
+                                            --                      4 is back
+        local cmd = code - (layerCode * 10) -- extract the actual "command"
+                                            -- 0-7 is a colour, 8 is xterm256
+        local colour = nil
+        if cmd == 8 and t[i+1] == '5' then -- xterm256, colour indexed
+          colour = convertindex(tonumber(t[i+2]))
+          i = i + 2
+
+        elseif cmd == 8 and t[i+1] == '2' then -- xterm256, rgb
+          colour = {t[i+2] or '', t[i+3] or '', t[i+4] or ''}
+          i = i + 4
```

Comment:
All right... We should be handling missing parameters (the `or ''`, which means, `if t[i+1] doesn't exist use the empty string`), though I am not sure we do the correct thing. Can decho handle empty parameters?

If parameters are left off, there can't be any other colour codes following, because otherwise, the codes become ambiguous: `\e38;2;12;42;48;2m` is the `38` the start of the next colour code or the actual value `48` for b? And since the index exceeds the length of the table, if no other values follow, we should be good here as well.

So all we may need to check is for empty strings as parts of `colour` (because the empty string is allowed as a parameter as well). I'll test that later.

@SlySven chime in if I missed something!